solr/core/src/java/org/apache/solr/cluster/placement/impl/CollectionMetricsBuilder.java

```diff
@@ -0,0 +1,117 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.cluster.placement.impl;
+
+import org.apache.solr.cluster.placement.CollectionMetrics;
+import org.apache.solr.cluster.placement.ReplicaMetrics;
+import org.apache.solr.cluster.placement.ShardMetrics;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ *
+ */
+public class CollectionMetricsBuilder {
+
+  final Map<String, ShardMetricsBuilder> shardMetricsBuilders = new HashMap<>();
+
+
+  public void addShardMetrics(String shardName, ShardMetricsBuilder shardMetricsBuilder) {
+    shardMetricsBuilders.put(shardName, shardMetricsBuilder);
+  }
+
+  public CollectionMetrics build() {
+    final Map<String, ShardMetrics> metricsMap = new HashMap<>();
+    shardMetricsBuilders.forEach((shard, builder) -> metricsMap.put(shard, builder.build()));
+    return shardName -> Optional.ofNullable(metricsMap.get(shardName));
+  }
+
+  public static class ShardMetricsBuilder {
+    final Map<String, ReplicaMetricsBuilder> replicaMetricsBuilders = new HashMap<>();
+
+    public ShardMetricsBuilder addReplicaMetrics(String replicaName, ReplicaMetricsBuilder replicaMetricsBuilder) {
+      replicaMetricsBuilders.put(replicaName, replicaMetricsBuilder);
+      return this;
+    }
+
+    public ShardMetricsBuilder setLeaderMetrics(ReplicaMetricsBuilder replicaMetricsBuilder) {
+      replicaMetricsBuilders.put(LEADER, replicaMetricsBuilder);
+      return this;
+    }
+
+    public static final String LEADER = "__leader__";
+
+    public ShardMetrics build() {
+      final Map<String, ReplicaMetrics> metricsMap = new HashMap<>();
+      replicaMetricsBuilders.forEach((name, replicaBuilder) -> {
+        ReplicaMetrics metrics = replicaBuilder.build();
+        metricsMap.put(name, metrics);
+        if (replicaBuilder.leader) {
+          metricsMap.put(LEADER, metrics);
+        }
+      });
+      return new ShardMetrics() {
+        @Override
+        public Optional<ReplicaMetrics> getLeaderMetrics() {
+          return Optional.ofNullable(metricsMap.get(LEADER));
+        }
+
+        @Override
+        public Optional<ReplicaMetrics> getReplicaMetrics(String replicaName) {
+          return Optional.ofNullable(metricsMap.get(replicaName));
+        }
+      };
+    }
+  }
+
+  public static class ReplicaMetricsBuilder {
+    final Map<String, Object> metrics = new HashMap<>();
+    int sizeGB = 0;
```

Comment:
Maybe use `Integer` here? Replicas always have size, even new ones, but if we can return null then we can signal that the size is unknown (eg. couldn't be retrieved from the node).